csrc/scheduler/mma_utils.cpp

```diff
@@ -1100,105 +1102,103 @@ ProblemIterDomainsOpt getProblemIterDomains(Fusion* fusion) {
-    Fusion* fusion,
-    const mma_utils::MulSumProperties::InputsOutputs& props) {
-  ComputeAtMap ca_map(fusion);
-  const auto mma_input_candidates =
-      ir_utils::filterByType<TensorView>(fusion->inputs()).vector();
-  if (mma_input_candidates.empty()) {
-    return {"Failed to find any TV that is fusion input"};
-  }
-
-  const auto mma_output_domains = getProblemIterDomains(props);
-  if (!mma_output_domains.isValid()) {
-    return mma_output_domains.getErrorMsg();
+MatmulProblemLayoutOpt getProblemLayout(Fusion* fusion) {
+  const std::vector<MatmulPattern> patterns = findMatmulPatterns(fusion);
+  if (patterns.size() != 1) {
+    std::stringstream ss;
+    ss << "Invalid number of MmaOp instances in fusion, expected 1, got "
+       << patterns.size();
+    return ss.str();
   }
+  return getProblemLayout(fusion, patterns.front());
+}
 
-  const auto domains_data = mma_output_domains.getData();
-  const auto m = domains_data[(size_t)MatmulDomain::M];
-  const auto n = domains_data[(size_t)MatmulDomain::N];
-  const auto k = domains_data[(size_t)MatmulDomain::K];
-
-  DependenciesMap deps_map;
-  resolveTvToMatmulDomainsMapping(
-      deps_map, mma_input_candidates, m, n, k, ca_map);
-
-  bool mk_found = false;
-  bool km_found = false;
-  bool nk_found = false;
-  bool kn_found = false;
-  const static DomainsDesc mk_desc = {MatmulDomain::M, MatmulDomain::K};
-  const static DomainsDesc km_desc = {MatmulDomain::K, MatmulDomain::M};
-  const static DomainsDesc nk_desc = {MatmulDomain::N, MatmulDomain::K};
-  const static DomainsDesc kn_desc = {MatmulDomain::K, MatmulDomain::N};
-
-  for (const auto& item : deps_map) {
-    if (item.second == mk_desc) {
-      if (mk_found) {
-        return {
-            "Failed to find MMA input, more than one fusion input has [..., M, ..., K, ...] iter domains"};
-      }
-      mk_found = true;
+MatmulProblemLayoutOpt getProblemLayout(
+    const IdModel& id_model,
+    const std::unordered_map<ValGroup, MatmulDomain>& group_to_domain,
+    const RolesMap& roles_map) {
+  // Assumes the exact graph has already been built, since we've been provided
+  // group_to_domain
+  const ValGraph& exact_graph = id_model.idGraph(IdMappingMode::EXACT);
+
+  // Note: using DataWrapperOpt<MatmulDomain> would be preferable here. However,
+  // using DataWrapperOpt<MatmulDomain>(std::move(dom)) leads to a clang-tidy
+  // warning because MatmulDomain is trivially movable. There is only a move
+  // constructor for DataWrapperOpt to prevent inadvertent copying. To avoid
+  // this complication I'm using a simple pair for the lambda's result type.
+  using InnerDomResult = std::pair<MatmulDomain, std::string>;
+  const auto innerDomain = [&roles_map, &group_to_domain, &exact_graph](
+                               MatmulRole role) -> InnerDomResult {
+    const auto role_it = roles_map.find(role);
+    if (role_it == roles_map.end()) {
+      return {MatmulDomain::M, "Could not find role in roles_map"};
```

Comment:
You're right and this `InnerDomResult` business is a hack to get around the design of `DataWrapperOpt`. I want to just return the error message in this case, but using `DataWrapperOpt` doesn't work properly in clang (haven't tried gcc) when the wrapped type is trivially copyable, since it balks at using `std::move` on such types.

Reply:
I just pushed a change that uses a bare `variant<std::string, UnitDim>`.